File: solutions/055.jump-game/jump-game.java
class Solution {
    public boolean canJump(int[] nums) {
        boolean canReach = true; 
        for(int i = 0; i < nums.length - 1; i++){
            if(nums[i] == 0){
                canReach = false; // 这个时候可能无法达到终点
                int j = i - 1;
                while(j>=0){
                    if(nums[j]-nums[i] > i - j){
                        canReach = true;
                        break;
                    }
                    j--;
                }
                if(!canReach){
                    break;
                }
            }
        }
        return canReach;
        
    }
}

File: solutions/055.jump-game/jump-game.js
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var canJump = function(nums) {
    let index = 0;
    let cnt = nums[0];
    while(cnt > 0 && index < nums.length - 1){
        let tempIndex = index;
        for(let i = 0; i <= cnt; i++){
            if(i + nums[index + i] > cnt){
                tempIndex += i;
                break;
            }
        }
        if(tempIndex === index){
            index += cnt;
            cnt =0;
        }else{
            cnt = nums[tempIndex];
            index = tempIndex;
        }
        
    }
    return index >= nums.length - 1
};